src/Commands/Images/GenerateAbstract.php

```diff
@@ -0,0 +1,376 @@
+                if (file_exists($dir . $imageObj->getImgFolder())) {
+                    $toDel = scandir($dir . $imageObj->getImgFolder(), SCANDIR_SORT_NONE);
+                    foreach ($toDel as $d) {
+                        foreach ($type as $imageType) {
+                            if (preg_match('/^[0-9]+\-' . $imageType['name'] . '\.jpg$/', $d) || (count($type) > 1 && preg_match('/^[0-9]+\-[_a-zA-Z0-9-]*\.jpg$/', $d))) {
+                                if (file_exists($dir . $imageObj->getImgFolder() . $d)) {
+                                    unlink($dir . $imageObj->getImgFolder() . $d);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Regenerate images.
+     *
+     * @param $dir
+     * @param $type
+     * @param bool $productsImages
+     *
+     * @return bool|string
+     */
+    protected function _regenerateNewImages($dir, $type, $productsImages = false)
+    {
+        $processTypes = [];
+        array_map(function ($row) use (&$processTypes) {
+            $processTypes[] = $row['name'];
+        }, $type);
+
+        $this->output->writeln('Processing formats ' . implode(', ', $processTypes));
+
+        if (!is_dir($dir)) {
+            $this->errors[] = 'Unable to regenerate new images for type ' . implode(',', $processTypes) . ' : ' . $dir . ' does not exists';
+
+            return false;
+        }
+
+        $generate_hight_dpi_images = (bool) Configuration::get('PS_HIGHT_DPI');
+
+        if (!$productsImages) {
+            $formated_medium = ImageType::getFormattedName('medium');
+            foreach (scandir($dir, SCANDIR_SORT_NONE) as $image) {
+                if (preg_match('/^[0-9]*\.jpg$/', $image)) {
+                    foreach ($type as $k => $imageType) {
+                        // Customizable writing dir
+                        $newDir = $dir;
+                        if (!file_exists($newDir)) {
+                            continue;
+                        }
+
+                        if (($dir == _PS_CAT_IMG_DIR_) && ($imageType['name'] == $formated_medium) && is_file(_PS_CAT_IMG_DIR_ . str_replace('.', '_thumb.', $image))) {
+                            $image = str_replace('.', '_thumb.', $image);
+                        }
+
+                        if (!file_exists($newDir . substr($image, 0, -4) . '-' . stripslashes($imageType['name']) . '.jpg')) {
+                            if (!file_exists($dir . $image) || !filesize($dir . $image)) {
+                                $this->errors[] = sprintf('Source file does not exist or is empty (%s)', $dir . $image);
+                            } elseif (!ImageManager::resize($dir . $image, $newDir . substr(str_replace('_thumb.', '.', $image), 0, -4) . '-' . stripslashes($imageType['name']) . '.jpg', (int) $imageType['width'], (int) $imageType['height'])) {
```

Comment:
with `$generate_hight_dpi_images` == true,
this line is executed, image generated, and so on line 268.
Image is generated 2 times.

Reply:
yes with this option the image file is generated twice.
One time with the normal format name and one time suffixed with 2x

Reply:
I think you can fix this before merging (not a hard job ;) )

Reply:
In fact this is not a bug , but a feature :)
In old themes to optimize the mobile image quality the image are replaced by the images with 2x 
``function highdpiInit()
{
	if (typeof highDPI === 'undefined')
		return;
	if(highDPI && $('.replace-2x').css('font-size') == "1px")
	{
		var els = $("img.replace-2x").get();
		for(var i = 0; i < els.length; i++)
		{
			src = els[i].src;
			extension = src.substr( (src.lastIndexOf('.') +1) );
			src = src.replace("." + extension, "2x." + extension);

			var img = new Image();
			img.src = src;
			img.height != 0 ? els[i].src = src : els[i].src = els[i].src;
		}
	}
}``
i don't know if it is still the case for some ps 1.7 themes

Reply:
Oh. I did not know.
Sorry for time waste.

Reply:
Retina displays can use 2x or 3x versions but it seems that PS_HIGHT_DPI is always 0 by default. I've checked an 1.5 and 1.7 real sites. I aven't found the highdpiInit js in any site.
So we should keep this code for now.

Reply:
I know this behavior as i'm using this feature on one site to improve image quality :D